### PR TITLE
Enable armv8 build using clang

### PR DESCRIPTION
**Pull request by anghelcovici**
### Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [X] Pull request includes prototype/experimental work that is under
      construction.

I've added armv8 as a target for clang. I've used this to build a arm binary and run it on Mendel linux on a Coral dev board. 

While doing this I've also done a small refactor of the build options by moving the ones that we should be able to play with  to the .bazelrc file.

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -31,11 +31,38 @@ build:wasm32 --symlink_prefix=bazel-wasm-
...
+# Simple but useful static checking to detect potential race conditions.
+build:clang --cxxopt="-Wthread-safety"
+# Treat warnings as errors.
+build:clang --cxxopt="-Werror"
+# Disable some warnings:
+# protobuf generates unusused parameters, so skip this for now.
+build:clang --cxxopt="-Wno-unused-parameter"
+# https://stackoverflow.com/questions/1538943
+build:clang --cxxopt="-Wno-missing-field-initializers"
+# Latest asylo update
```
Comment (tiziano88) [inline .bazelrc]: s/asylo/Asylo/

File: .bazelrc [hunk 2]
```diff
@@ -31,11 +31,38 @@ build:wasm32 --symlink_prefix=bazel-wasm-
...
+# Enable all warnings.
+build:clang --cxxopt="-Wall"
+# Enable extra warnings.
+build:clang --cxxopt="-Wextra"
+# Simple but useful static checking to detect potential race conditions.
+build:clang --cxxopt="-Wthread-safety"
+# Treat warnings as errors.
+build:clang --cxxopt="-Werror"
+# Disable some warnings:
+# protobuf generates unusused parameters, so skip this for now.
```
Comment (daviddrysdale) [inline .bazelrc]: s/unusused/unused/

File: .bazelrc [hunk 3]
```diff
@@ -31,11 +31,38 @@ build:wasm32 --symlink_prefix=bazel-wasm-
...
+build:clang --cxxopt="-fcolor-diagnostics"
+# Don't omit the frame pointer, useful for debug.
+build:clang --cxxopt="-fno-omit-frame-pointer"
+# Enable all warnings.
+build:clang --cxxopt="-Wall"
+# Enable extra warnings.
+build:clang --cxxopt="-Wextra"
+# Simple but useful static checking to detect potential race conditions.
+build:clang --cxxopt="-Wthread-safety"
+# Treat warnings as errors.
```
Comment (daviddrysdale) [inline .bazelrc]: How [fragile](https://embeddedartistry.com/blog/2017/05/22/werror-is-not-your-friend/) is this?  If we happen to update a dependency that has a new warning, or if we upgrade Clang version to one with more warnings on existing code, isn't this going to break the build?

File: oak/server/oak_runtime.cc [hunk 4]
```diff
@@ -158,20 +156,6 @@ bool OakRuntime::CreateAndRunNode(const std::string& config_name,
 }
 
 grpc::Status OakRuntime::Start() {
-  // We call into the Rust runtime to verify that bindings between C++ and Rust are working
-  // correctly.
```
Comment (daviddrysdale) [inline oak/server/oak_runtime.cc]: Obviously this was just a placeholder, but I presume @blaxill's work on getting a live Rust runtime is going to clash with this?

File: oak/server/oak_runtime.cc [hunk 5]
```diff
@@ -158,20 +156,6 @@ bool OakRuntime::CreateAndRunNode(const std::string& config_name,
 }
 
 grpc::Status OakRuntime::Start() {
-  // We call into the Rust runtime to verify that bindings between C++ and Rust are working
```
Comment (ipetr0v) [inline oak/server/oak_runtime.cc]: Why is this code being deleted?
Are we deprecating the Rust Runtime?

File: toolchain/cc_toolchain_config.bzl [hunk 6]
```diff
@@ -103,19 +103,20 @@ def _impl(ctx):
         ),
     ]
 
-    # System include path we need to setup for the downloaded clang
-    isystem_flags = feature(
-        name = "isystem_flags",
```
Comment (ipetr0v) [inline toolchain/cc_toolchain_config.bzl]: Aren't `isystem_flags` necessary for `k8` ?

File: WORKSPACE [hunk 7]
```diff
@@ -323,12 +323,21 @@ go_rules_compat(
...
 http_archive(
     name = "clang",
-    build_file = "//toolchain:clang.BUILD",
+    build_file = "//toolchain:all_files.BUILD",
     sha256 = "0f5c314f375ebd5c35b8c1d5e5b161d9efaeff0523bac287f8b4e5b751272f51",
     strip_prefix = "clang+llvm-8.0.0-x86_64-linux-gnu-ubuntu-18.04",
     url = "http://releases.llvm.org/8.0.0/clang+llvm-8.0.0-x86_64-linux-gnu-ubuntu-18.04.tar.xz",
 )
 
+# sysroot for arm
```
Comment (daviddrysdale) [inline WORKSPACE]: Could you expand this comment?  Is this a host=linux target=arm compiler?

File: toolchain/BUILD [hunk 8]
```diff
@@ -62,6 +68,16 @@ filegroup(
     ],
 )
 
+filegroup(
```
Comment (daviddrysdale) [inline toolchain/BUILD]: Could do with some comments explaining what/why.

File: toolchain/cc_toolchain_config.bzl [hunk 9]
```diff
@@ -136,6 +159,7 @@ def _impl(ctx):
                     flag_group(
                         flags = [
                             "-lstdc++",
+                            "-fuse-ld=lld",
```
Comment (ipetr0v) [inline toolchain/cc_toolchain_config.bzl]: Why do we need `-fuse-ld` in `k8` ?

Comment (anghelcovici) [reply toolchain/cc_toolchain_config.bzl]: We don't... It's faster :)

File: oak/server/oak_runtime.cc (hunk 4 above)
Comment (anghelcovici) [reply oak/server/oak_runtime.cc]: I'll let @blaxill comment on this. He told me it is safe to remove.... IIRC :)

Comment (blaxill) [reply oak/server/oak_runtime.cc]: This is safe to remove. It was to check that we could compile no_std rust code into Asylo. Originally added by tzn, I continued to use it for asylo_rust code. The first iteration of rust oak_runtime I'm working on currently does not use asylo_rust but will in the future.

File: oak/server/oak_runtime.cc (hunk 5 above)
Comment (anghelcovici) [reply oak/server/oak_runtime.cc]: No, please see @blaxill's comment below.

File: toolchain/cc_toolchain_config.bzl (hunk 6 above)
Comment (anghelcovici) [reply toolchain/cc_toolchain_config.bzl]: No. This is a bit of a hack for now, but we don't have a sysroot for local clang so we just use /usr/include. We need to have a Linux x64 sysroot at some point, if we want to make the dev server properly platform independent.

File: .bazelrc (hunk 3 above)
Comment (anghelcovici) [reply .bazelrc]: Talked about this. Let's keep it like this for a bit and if it causes issues we drop it.

File: toolchain/BUILD (hunk 8 above)
Comment (anghelcovici) [reply toolchain/BUILD]: Done

File: WORKSPACE (hunk 7 above)
Comment (anghelcovici) [reply WORKSPACE]: Done

File: .bazelrc (hunk 2 above)
Comment (anghelcovici) [reply .bazelrc]: Done

File: .bazelrc (hunk 1 above)
Comment (anghelcovici) [reply .bazelrc]: Done

**Pull request merged**